### DISCUSSION TITLE
Make the attribute's qualifier separator configurable (the dot breaks JSX)?

### DISCUSSION THREAD

**Discussion by claudioc** [alpinejs/alpine#4164] [3. Feature ideas]
I am building an application that is (only) rendered on the server (Fastify); I am building it in the open, so you can [take a look](https://github.com/claudioc/joongle). I am using JSX for server-side templating and Preact as the server-side JSX renderer. I also use Alpine and HTMX for client-side interactivity. So far, so good - the project is a BIG proof-of-concept of this idea of using mostly SSR but with modern tools, like JSX and reactivity via Alpine, without the need of a bundler...

The issue is that I cannot use qualifiers in Alpine's attribute, as in `x-on:keydown.escape` since the dot is not syntactically valid for Preact (also Prettier and eslint apparently).

I wonder if we could somehow configure that dot to be another character, like "_" (which is probably safe, in Alpine's context).

What do you think? Is there any other way to fix this problem that I haven't thought of?

Attaching a screenshot just for reference.

Thank you!

![Screenshot 2024-04-21 at 15 01 25](https://github.com/alpinejs/alpine/assets/166692/2cb45761-6fa8-4bdd-bada-61c0c5fe4f98)

**Comment by ekwoka** [comment]
I believe _ is not allowed in attribute names. And . Is tough in jsx.

Have you thought about using astro for this instead of Preact? It works nicely there.

**Reply by claudioc** [reply]
Astro is really cool, but as I am building a "product" more than a single instance of website/webapp, I want to have as much as flexibility in the tech stack as possible. Also, the issue with Alpine+JSX is not really a blocker - it's just a small annoying thing at the moment

Preact here is only used to compile the JSX on the server (there is no Preact running on the client)

**Comment by ekwoka** [comment]
Checked, and `_` is valid in html attributes.

I feel like it should be possible to make this work with your linters.

Does this not work at all? or just TS complains?

**Reply by claudioc** [reply]
Besides the linters, I also tried with just both TS and esbuild (which I use), see screenshots. They both fail with a similar error. I may try another "JSX compiler" instead of Preact - but that would be React :/

<img width="522" alt="Screenshot 2024-04-22 at 13 00 29" src="https://github.com/alpinejs/alpine/assets/166692/ec5d7d2c-0636-4f9d-8719-c0d8c8c4bb87">
<img width="729" alt="Screenshot 2024-04-22 at 12 58 11" src="https://github.com/alpinejs/alpine/assets/166692/66841ab7-b107-4535-be5a-8ce4e8881f81">

**Comment by claudioc** [comment]
In the meantime, I think I have found a workaround - using the spread operator

```jsx
  const onKeypress = {
    'x-on:keypress.escape': 'alert(123)',
  };

  return (
    <div {...onKeypress}>
      {/* something something */}
    </div>
  );
```

**Comment by WhiskeyTuesday** [comment]
Another way to do the same thing, also a bit of a crusty hack but in my opinion maybe less so.

```jsx
    <div { ...{
        'x:on:keypress.escape': 'alert(123)',
        'x:otherattribute': 'whatever()',
     } }>
      {/* something something */}
    </div>
  );
